### PR TITLE
allow synchronizing user status from OAuth2 login providers

### PR DESCRIPTION
**Pull request by bohde**
This leverages the existing `sync_external_users` cron job to synchronize the `IsActive` flag on users who use an OAuth2 provider set to synchronize. This synchronization is done by checking for expired access tokens, and using the stored refresh token to request a new access token. If the response back from the OAuth2 provider is the `invalid_grant` error code, the user is marked as inactive. However, the user is able to reactivate their account by logging in the web browser through their OAuth2 flow.

Also changed to support this is that a linked `ExternalLoginUser` is always created upon a login or signup via OAuth2.

### Notes on updating permissions
Ideally, we would also refresh permissions from the configured OAuth provider (e.g., admin, restricted and group mappings) to match the implementation of LDAP. However, the OAuth library used for this `goth`, doesn't seem to support issuing a session via refresh tokens. The interface provides a [`RefreshToken` method](https://github.com/markbates/goth/blob/master/provider.go#L20), but the returned `oauth.Token` doesn't implement the `goth.Session` we would need to call `FetchUser`. Due to specific implementations, we would need to build a compatibility function for every provider, since they cast to concrete types (e.g. [Azure](https://github.com/markbates/goth/blob/master/providers/azureadv2/azureadv2.go#L132))

### CODE REVIEW

**Comment by kdumontnu** [comment]
Backlinking to tracking issue https://github.com/go-gitea/gitea/issues/23794

File: models/user/external_login_user.go [hunk 1]
```diff
@@ -160,12 +160,36 @@ func UpdateExternalUserByExternalID(ctx context.Context, external *ExternalLogin
...
-	Provider string
-	UserID   int64
-	OrderBy  string
+	Provider        string
+	UserID          int64
+	LoginSourceID   int64
+	HasRefreshToken bool
+	Expired         bool
+	OrderBy         string
+	Limit           int
```
Comment (6543) [inline models/user/external_login_user.go]: why do you have Limit and Start here ... this should be covered by 	db.ListOptions ?!?

File: models/user/external_login_user.go [hunk 2]
```diff
@@ -160,12 +160,34 @@ func UpdateExternalUserByExternalID(ctx context.Context, external *ExternalLogin
 	return err
 }
 
+// EnsureLinkExternalToUser link the external user to the user
+func EnsureLinkExternalToUser(external *ExternalLoginUser) error {
```
Comment (6543) [inline models/user/external_login_user.go]: Instead of db.DefaultContext could you pass the ctx throug?

Comment (6543) [reply models/user/external_login_user.go]: Yes confirmed this could case a deadlock!

File: models/user/external_login_user.go (hunk 1 above)
Comment (bohde) [reply models/user/external_login_user.go]: Thanks! It looks like I accidentally kept this in during merge conflict resolution. Fixed in https://github.com/go-gitea/gitea/pull/31572/commits/0327ff2bf11296f14fbf06d98810f0565af19843

File: models/user/external_login_user.go (hunk 2 above)
Comment (bohde) [reply models/user/external_login_user.go]: Fixed in https://github.com/go-gitea/gitea/pull/31572/commits/2b71ec7a225b215632e498306060870de6295451

**Review by 6543 — APPROVED** [review]

**Review by kdumontnu — APPROVED** [review]
👍

**Pull request merged**